### ISSUE TITLE
delete autogenerated bindings to platform interface

### ISSUE THREAD

**Issue by mrhota** [ogham/rust-locale#9]
1. Platform bindings live in rust-lang/libc.
2. The binding generator naively created constants for all internal langinfo #defines. Most of that stuff shouldn't ever be seen by users. That's why they start with `_`...

Trying to deal with GNU's whimsical crap is insanely complicated. Let's K.I.S.S.

**Comment by jan-hudec** [comment]
I have them unpublished locally already.

I don't agree trying to deal with _GNU's_ crap being insanely complicated though. The `grouping` and `mon_grouping` fields in [struct localeconv](http://pubs.opengroup.org/onlinepubs/9699919799/functions/localeconv.html) have the same exceptions. GNU LibC is better than other LibCs in that it at least provides _everything_ it knows via the `nl_langinfo` function.

**Comment by jan-hudec** [comment]
> Most of that stuff shouldn't ever be seen by users. That's why they start with `_`

Actually no; I am pretty sure that stuff _is_ intended to be seen by users. They start with `_` because they are extension and the standards actually don't allow system headers to define extensions that don't start with `_`.

It is, however, system specific, so we'll need to provide it ourselves for the systems that don't have it anyway. And when we do, there is no reason to use it on GNU either.

**Comment by mrhota** [comment]
I think it's still not our problem. People who want or need `grouping` and `mon_grouping` can handle it themselves using rust-lang/libc. Our library will provide a more complete, more sane interface than platform libc. Let's not pollute this library with it.

**Comment by jan-hudec** [comment]
No, we definitely shouldn't be exporting any bindings to C runtime. And we won't.

**Comment by mrhota** [comment]
Agreed. Can we agree that we shouldn't _wrap_ platform interfaces either?

**Comment by jan-hudec** [comment]
We won't.

We definitely won't do so publically.

When the https://github.com/rust-lang/libc/pull/248 is merged, we won't need to be binding any functions internally either. We will still need to do the interpretation in `localeconv` and possibly define some of the system-specific constants.

However I am somewhat leaning towards _first_ doing bundled data taken from CLDR and then, well, there probably won't be much reason to implement any kind of alternate backend reading stuff from libc.

**Comment by mrhota** [comment]
> We will still need to do the interpretation in `localeconv` and possibly define some of the system-specific constants.

I can't imagine a reason we would care what `localeconv` has to say about anything. Maybe I don't understand what you mean, though. Can you expand on why you think `localeconv` should/would be used in rust-locale? Imagine we're using CLDR data _only_. Does that change things?

If you look at how the java runtime uses platform locale functions [when it boots](http://hg.openjdk.java.net/jdk9/jdk9/jdk/file/edb95a70985f/src/java.base/unix/native/libjava/java_props_md.c#l98), you'll see it basically ignores everything except `setlocale`, and it only does that to figure out what language tag to use so it can set JVM system properties (ok, it also calls `nl_langinfo(CODESET)` :smile:). When I said (in #8) rust-locale should call `setlocale` once during init, I meant that it should be used to determine the system's current locale. Other than that, I say we don't care (for now) what the platform has to say about locale stuff.

If we want to get our initial bearings on a system by calling `setlocale`, we take the language tag it returns and use the associated data from CLDR.

**Comment by mrhota** [comment]
> leaning towards first doing bundled data taken from CLDR

_Absolutely!_ There are a variety of ways that we can distribute 1) all CLDR data or 2) a subset that the user specifies. I'm thinking [cargo [features]](http://doc.crates.io/manifest.html#the-features-section) together with a CLDR parser, similar in spirit to python-babel's, but there might be better ways. Anyway, the details don't matter right now. I'll create an issue for it.

**Comment by mrhota** [comment]
See #10

**Comment by jan-hudec** [comment]
> I can't imagine a reason we would care what localeconv has to say about anything.

Only if we want to use the system data for anything. If we just use CLDR, there is indeed no point.

I was thinking about using the system data, or using them initially, but it will probably be easier to just use CLDR. In which case, honestly, I don't see point in using _any_ libc locale function, _ever_. Because

> If you look at how the java runtime uses platform locale functions **when it boots**

That's exactly where we can't call it, because Rust does not have global constructors. So the only way we can call it is lazy and I fear breaking stuff by calling it at random point.

Which is why I am thinking about getting the value from the environment variables directly. The way GNU LibC uses them is documented and can be checked in source, Musl won't differ, I don't think OSX will either and Windows have to be special-cased anyway, since they use different names, so we can just go and use the native `GetUserDefaultLCID`.

> Other than that, I say we don't care (for now) what the platform has to say about locale stuff.

Well, perhaps we, one day, should, because there are methods for users to define their own locales.

Also, I originally thought about first using the libc data and only then going on to extract them from CLDR. But since extracting from CLDR covers all systems at once, I now agree that going directly for that is the better approach.

**Comment by mrhota** [comment]
> the only way we can call it is lazy

You're right. ~~Imagine a method like `Locale::getDefaultLocale()`. If not already constructed, that's when `setlocale` could be called~~ Actually, you're right, what's the point? We should determine platform locale from the same place libc does. Like you said, env. variables, or even some config file we can specify.

I'm glad we're hashing this out. Looks like we're on the same page :smiley: 

> perhaps we, one day, should

Yes. Definitely something to consider for the future.

**Comment by jan-hudec** [comment]
Note: this is part of 0.4 and not 0.3 only so that we get a release cycle with deprecation.
